src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataFilter.cs

```diff
@@ -82,14 +106,25 @@ public void OnResultExecuted(ResultExecutedContext context)
             if (!context.HttpContext.Response.HasStarted)
             {
                 SaveTempData(context.Result, _factory, context.Filters, context.HttpContext);
-                // If SaveTempDataFilter got added twice this might already be in there.
-                if (!context.HttpContext.Items.ContainsKey(TempDataSavedKey))
+
+                var saveTempDataContext = GetTempDataContext(context.HttpContext);
+                if (saveTempDataContext != null)
                 {
-                    context.HttpContext.Items.Add(TempDataSavedKey, true);
+                    saveTempDataContext.TempDataSaved = true;
                 }
             }
         }
 
+        private SaveTempDataContext GetTempDataContext(HttpContext httpContext)
+        {
+            SaveTempDataContext saveTempDataContext = null;
+            if (httpContext.Items.TryGetValue(SaveTempDataFilterContextKey, out var value))
+            {
+                saveTempDataContext = value as SaveTempDataContext;
```

Comment:
Should this be a regular cast? I think it would be fine to blow up if something \ someone replaced it with a random object instance.